### DISCUSSION TITLE
Internal temperature sensor not accessible on Pico 2?

### DISCUSSION THREAD

**Discussion by ChrisHinde** [micropython/micropython#15732] [RP2040 / Pico]
I'm just playing around and tried to read the internal temperature sensor ("core temp"), but I can't get it to work (in serial console).
When I try setting it up I get an error saying that the pin doesn't have ADC capabilities.

The example I started with is found here: https://github.com/raspberrypi/pico-micropython-examples/blob/master/adc/temperature.py (as well as in some docs).
But I've also looked at RP2350 datasheet to confirm that it's the right channel, etc.

Here's some examples of what I've tried (removed Traceback for brevity):

```
>>> sensor_temp = machine.ADC(4)
ValueError: Pin doesn't have ADC capabilities
>>> adc = ADC(ADC.CORE_TEMP)
ValueError: Pin doesn't have ADC capabilities
```
ADC.CORE_TEMP does return 4

I can setup and read from other ADCs
```
>>> adc = ADC(0)
>>> adc.read_u16()
11522
>>> adc = ADC(29)
>>> adc.read_u16()
30135
```

`MicroPython v1.24.0-preview.201.g269a0e0e1 on 2024-08-09; Raspberry Pi Pico2 with RP2350-RISCV`

Also tried on

`MicroPython v1.24.0-preview.201.g269a0e0e1 on 2024-08-09; Raspberry Pi Pico2 with RP2350`

Is there a trick I'm missing, is it a bug/not yet supported (I know that there are no official builds etc yet)??

**Comment by rikkenharp** [comment]
Hi. 
Look into my Gist.
I'm reading directly from the register.
https://gist.github.com/rikkenharp/06e62960df8f23e9da748edc625a6215
Oddly pin 3 has ADC capabilities or is that ok?

**Reply by ChrisHinde** [reply]
I saw that, my main concern is about accessing documented features in "documented"/expect ways.
I weren't really interested in reading the temperature, just messing around. :)

I tried your code, and it seems to work. (might be useful in the future, but after reading another pin, it spits out the same value as the previous reading, on the first run of the function)
As for "pin 3", that would be ADC3 or GPIO29 (or pin 43)‽ It's not broken out on the Pico 2 board, but is used to read the ADC_VREF

**Comment by robert-hh** [comment]
Small numbers like are the ADC channel numbers, Larger numbers like 28 are GPIO numbers. See https://docs.micropython.org/en/latest/rp2/quickref.html#adc-analog-to-digital-conversion

**Reply by ChrisHinde** [reply]
I know.

But my issue is that I can setup, and read, from ADC0-3, but not from ADC4 (internal temp., and does not have an GPIO pin associated), as per 12.4.2.1 in the datasheet: 
![Skärmbild 2024-08-29 214213](https://github.com/user-attachments/assets/6837ea06-de9f-4fd9-834c-72be35ff8370)
(The error message saying "Pin" is perhaps not entirely correct, but I guess: 1. It defaults to taking the "ID" as a pin and not an ADC channel. 2. More specific error messages means more resources etc)

Now, I know that this is an early build, and that most of the code is basically the same as for "Pico 1"/RP2040, so I guess there some differences in the ADC setup between them.

Looking at the source code for the ADC, it does seem like it should work.

**Comment by ChrisHinde** [comment]
I now see an issue reported on this (#15687), I don't know how I missed it. I did do a search, but it didn't come up with any reported issues.
